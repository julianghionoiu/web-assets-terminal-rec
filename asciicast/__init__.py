import json
from decimal import Decimal

from colorama import Fore, Style


class Text:
    NEWLINE = "\r\n"
    CONSOLE_ARROW = Fore.CYAN + "âžœ " + Style.RESET_ALL


def wait(ticks):
    return [{"duration_in_ticks": ticks, "content": ""}]


def appear(content):
    return [{"duration_in_ticks": 1, "content": content}]


def typing(text="hostname"):
    return [{"duration_in_ticks": 1, "content": char} for char in text]


def delete(num_chars):
    return [{"duration_in_ticks": 1, "content": "\b \b"} for _ in range(0, num_chars)]


def press_enter():
    return [{"duration_in_ticks": 1, "content": Text.NEWLINE}]


def red(text):
    return Fore.RED + text + Style.RESET_ALL


def green(text):
    return Fore.GREEN + text + Style.RESET_ALL


def dim(text):
    return Style.DIM + text + Style.RESET_ALL


def bright(text):
    return Style.BRIGHT + text + Style.RESET_ALL


class DecimalEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, Decimal):
            return float("{:.2f}".format(obj))
        # Let the base class default method raise the TypeError
        return json.JSONEncoder.default(self, obj)


def write_asciicast_v2(stream, metadata, asciicast_lines):
    length_of_one_tick_in_seconds = 0.10
    current_time_sec = Decimal(0.00)
    json.dump(metadata, stream)
    stream.write("\n")
    for line in asciicast_lines:
        asciicast_v2_line = [current_time_sec, "o", line["content"]]
        current_time_sec += Decimal(length_of_one_tick_in_seconds * line["duration_in_ticks"])

        json.dump(asciicast_v2_line, stream, cls=DecimalEncoder)
        stream.write("\n")
