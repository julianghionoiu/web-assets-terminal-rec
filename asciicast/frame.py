class AsciicastFrame(object):
    duration_in_ticks: int
    content: str

    def __init__(self, duration_in_ticks: int, content: str) -> None:
        self.duration_in_ticks = duration_in_ticks
        self.content = content
