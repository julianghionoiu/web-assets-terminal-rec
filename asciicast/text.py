def console_arrow() -> str:
    return "\u001b[36mâžœ \u001b[m"


def newline() -> str:
    return "\r\n"


# ~~~~~~~~~~ Colours

def red(text: str) -> str:
    return "\u001b[31m{}\u001b[0m".format(text)


def green(text: str) -> str:
    return "\u001b[32m{}\u001b[0m".format(text)


def dim(text: str) -> str:
    return "\u001b[2m{}\u001b[0m".format(text)


def bright(text: str) -> str:
    return "\u001b[1m{}\u001b[0m".format(text)


# ~~~~~~~~~~ Utils

def trim(text: str, length: int):
    return (text[:20] + '..') if len(text) > length else text


def replace_escapes(text: str):
    return text.replace("\u001b", "Esc")
