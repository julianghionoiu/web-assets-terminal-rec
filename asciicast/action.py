from abc import ABC, abstractmethod
from typing import List

from asciicast.frame import AsciicastFrame


class AsciicastAction(ABC):

    @abstractmethod
    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        ...
