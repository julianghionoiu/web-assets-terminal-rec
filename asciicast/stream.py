import json
from decimal import Decimal

from asciicast.frame import AsciicastFrame


class AsciicastStream(object):
    width: int
    height: int

    def __init__(self, width: int, height: int, stream, length_of_one_tick_in_seconds=0.10) -> None:
        self.width = width
        self.height = height
        self._length_of_one_tick_in_seconds = length_of_one_tick_in_seconds

        # Start the stream
        self._stream = stream
        self._current_time_sec = Decimal(0.00)
        metadata = {"version": 2, "width": self.width, "height": self.height}
        json.dump(metadata, stream)
        stream.write("\n")

    def write_frame(self, frame: AsciicastFrame):
        asciicast_v2_line = [self._current_time_sec, frame.frame_type, frame.content]
        self._current_time_sec += Decimal(self._length_of_one_tick_in_seconds * frame.duration_in_ticks)
        json.dump(asciicast_v2_line, self._stream, cls=DecimalEncoder)
        self._stream.write("\n")


class DecimalEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, Decimal):
            return float("{:.2f}".format(obj))
        # Let the base class default method raise the TypeError
        return json.JSONEncoder.default(self, obj)
