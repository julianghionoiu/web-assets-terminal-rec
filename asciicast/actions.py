from typing import List

from asciicast.action import AsciicastAction
from asciicast.frame import AsciicastFrame
from asciicast.text import newline


class Wait(AsciicastAction):
    _ticks: int

    def __init__(self, ticks: int) -> None:
        self._ticks = ticks

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(self._ticks, "")]


class Appear(AsciicastAction):
    _content: str

    def __init__(self, content: str) -> None:
        self._content = content

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, self._content)]


class TypeChars(AsciicastAction):
    _text: str

    def __init__(self, text: str) -> None:
        self._text = text

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, char) for char in self._text]


class Delete(AsciicastAction):
    _num_chars: int

    def __init__(self, num_chars: int) -> None:
        self._num_chars = num_chars

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, "\b \b") for _ in range(0, self._num_chars)]


class PressEnter(AsciicastAction):

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, newline())]
