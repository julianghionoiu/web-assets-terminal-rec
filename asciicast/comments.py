from asciicast.frame import AsciicastFrame, AsciicastFrameType
from asciicast.stream import AsciicastStream


class Comments(object):

    def __init__(self, asciicast_stream: AsciicastStream) -> None:
        self._asciicast_stream = asciicast_stream

    def add(self, comment: str) -> None:
        self._asciicast_stream.write_frame(
            AsciicastFrame(0,
                           frame_type=AsciicastFrameType.INPUT,
                           content=comment))
