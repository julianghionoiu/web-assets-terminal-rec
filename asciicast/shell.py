from asciicast.stream import AsciicastStream
from asciicast.text import newline, trim, replace_escapes


class Shell(object):

    def __init__(self, asciicast_stream: AsciicastStream) -> None:
        self._stream = asciicast_stream

    def wait(self, ticks: int) -> None:
        self._stream.write_section_comment("wait(ticks={})".format(ticks))
        self._stream.write_frame(ticks, content="")
        self._stream.write_frame(0, content="")

    def appear(self, content: str) -> None:
        self._stream.write_section_comment("appear(content={})".format(replace_escapes(trim(content, 20))))
        self._stream.write_frame(1, content=content)

    def type_chars(self, text: str) -> None:
        self._stream.write_section_comment("type_chars(text={})".format(text))
        for _char in text:
            self._stream.write_frame(1, content=_char)

    def delete(self, num_chars: int) -> None:
        self._stream.write_section_comment("delete(num_chars={})".format(num_chars))
        for _ in range(0, num_chars):
            self._stream.write_frame(1, content="\b \b")

    def press_enter(self):
        self._stream.write_frame(1, content=newline())
