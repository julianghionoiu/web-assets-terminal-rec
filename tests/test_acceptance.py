import io

from approvaltests.approvals import verify

from asciicast.shell import Shell
from asciicast.stream import AsciicastStream
from asciicast.text import newline, red, dim, bright, green
from asciicast.comments import Comments


def test_appear_and_type():
    string_stream = io.StringIO()
    shell = Shell(AsciicastStream(width=60, height=21, stream=string_stream))

    shell.wait(ticks=10)
    shell.appear("> ")
    shell.type_chars("hostname")
    shell.press_enter()
    shell.appear("candidate-laptop.local" + newline())

    verify(string_stream.getvalue())


def test_delete():
    string_stream = io.StringIO()
    shell = Shell(AsciicastStream(width=60, height=21, stream=string_stream))

    shell.appear("> ")
    shell.appear("foocar")
    shell.delete(3)
    shell.appear("bar")

    verify(string_stream.getvalue())


def test_colours():
    string_stream = io.StringIO()
    shell = Shell(AsciicastStream(width=60, height=21, stream=string_stream))

    shell.appear("1. " + red("some red text") + " " + green('some green text') + newline())
    shell.appear("2. " + dim(red('x')) + " " + bright(red('x')) + newline())

    verify(string_stream.getvalue())


def test_add_comments_to_stream():
    string_stream = io.StringIO()
    aciicast_stream = AsciicastStream(width=60, height=21, stream=string_stream)
    shell = Shell(aciicast_stream)
    comments = Comments(aciicast_stream)

    shell.wait(ticks=5)
    comments.add("This is a comment")
    shell.appear("candidate-laptop.local" + newline())

    verify(string_stream.getvalue())

# ~~~~~~ Test helper
