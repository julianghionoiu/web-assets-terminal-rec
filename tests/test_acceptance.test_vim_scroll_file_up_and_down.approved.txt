{"version": 2, "width": 60, "height": 21}
[0.0, "i", "~~~~~~~~~~~~~  display_content(block_of_text=line0\nline1\nline2\nli..)  ~~~~~~~~~~~~~"]
[0.0, "i", "## -- set duration"]
[0.0, "o", ""]
[0.1, "i", "## -- clear screen"]
[0.1, "o", "\u001b[?25l"]
[0.1, "o", "\u001b[1;21r"]
[0.1, "o", "\u001b[H"]
[0.1, "o", "\u001b[2J"]
[0.1, "i", "## -- draw screen"]
[0.1, "o", "line0\r\n"]
[0.1, "o", "line1\r\n"]
[0.1, "o", "line2\r\n"]
[0.1, "o", "line3\r\n"]
[0.1, "o", "line4\r\n"]
[0.1, "o", "line5\r\n"]
[0.1, "o", "line6\r\n"]
[0.1, "o", "line7\r\n"]
[0.1, "o", "line8\r\n"]
[0.1, "o", "line9\r\n"]
[0.1, "o", "line10\r\n"]
[0.1, "o", "line11\r\n"]
[0.1, "o", "line12\r\n"]
[0.1, "o", "line13\r\n"]
[0.1, "o", "line14\r\n"]
[0.1, "o", "line15\r\n"]
[0.1, "o", "line16\r\n"]
[0.1, "o", "line17\r\n"]
[0.1, "o", "line18\r\n"]
[0.1, "o", "line19\r\n"]
[0.1, "i", "## -- set and show cursor"]
[0.1, "o", "\u001b[H"]
[0.1, "o", "\u001b[?25h"]
[0.1, "i", "~~~~~~~~~~~~~  cursor_down(num_lines=5)  ~~~~~~~~~~~~~"]
[0.1, "i", "## -- lines_with_cursor_only_move=5"]
[0.1, "i", "## -- content_beyond_current_scroll=0"]
[0.1, "i", "## -- data_lines_not_yet_shown=5"]
[0.1, "i", "## -- lines_with_content_regen_scroll=0"]
[0.1, "i", "## -- set duration"]
[0.1, "o", ""]
[0.2, "o", "\u001b[5B"]
[0.2, "i", "~~~~~~~~~~~~~  cursor_down(num_lines=14)  ~~~~~~~~~~~~~"]
[0.2, "i", "## -- lines_with_cursor_only_move=14"]
[0.2, "i", "## -- content_beyond_current_scroll=0"]
[0.2, "i", "## -- data_lines_not_yet_shown=5"]
[0.2, "i", "## -- lines_with_content_regen_scroll=0"]
[0.2, "i", "## -- set duration"]
[0.2, "o", ""]
[0.3, "o", "\u001b[14B"]
[0.3, "i", "~~~~~~~~~~~~~  cursor_down(num_lines=5)  ~~~~~~~~~~~~~"]
[0.3, "i", "## -- lines_with_cursor_only_move=0"]
[0.3, "i", "## -- content_beyond_current_scroll=5"]
[0.3, "i", "## -- data_lines_not_yet_shown=5"]
[0.3, "i", "## -- lines_with_content_regen_scroll=5"]
[0.3, "i", "## -- set duration"]
[0.3, "o", ""]
[0.4, "i", "## -- push content up"]
[0.4, "o", "\u001b[?25l"]
[0.4, "o", "\u001b[1;20r"]
[0.4, "o", "\u001b[20;1H"]
[0.4, "o", "\r\n"]
[0.4, "o", "\r\n"]
[0.4, "o", "\r\n"]
[0.4, "o", "\r\n"]
[0.4, "o", "\r\n"]
[0.4, "i", "## -- add new lines"]
[0.4, "o", "\u001b[16;1H"]
[0.4, "o", "line20"]
[0.4, "o", "\u001b[17;1H"]
[0.4, "o", "line21"]
[0.4, "o", "\u001b[18;1H"]
[0.4, "o", "line22"]
[0.4, "o", "\u001b[19;1H"]
[0.4, "o", "line23"]
[0.4, "o", "\u001b[20;1H"]
[0.4, "o", "line24"]
[0.4, "i", "## -- reset original viewport + footer"]
[0.4, "o", "\u001b[1;21r"]
[0.4, "i", "## -- show cursor at the end"]
[0.4, "o", "\u001b[20;1H"]
[0.4, "o", "\u001b[?25h"]
[0.4, "i", "~~~~~~~~~~~~~  wait(ticks=10)  ~~~~~~~~~~~~~"]
[0.4, "o", ""]
[1.4, "o", ""]
